src/TypeInference/ExprUtils.ml

```diff
@@ -155,6 +155,11 @@ and instantiate_named_param ~nset ~inst env (e : T.expr) (name, isch) =
   | None, T.NVar x ->
     (* TODO: we could provide freshly bound parameters here *)
     Error.fatal (Error.unbound_named_param ~pos:e.pos x)
+  | None, T.NOptionalVar x ->
+    assert (T.Scheme.is_monomorphic isch);
+    let arg = PreludeTypes.mk_None ~env ~pos:e.pos isch.sch_body in
+    { T.pos = e.pos; T.data = T.EApp(e, arg) }
```

Comment:
This implementation suffers problem similar to #11: implicit arguments are not reinstatiated when function has polymorphic argument. Here is an example code that exhibit this behavior:
```
let foo (f : {?x:Int} -> Unit ->[] Int) = f {x=42} ()
let bar {?x} =
  match x with
  | None => 0
  | Some y => y
  end

let z = foo bar // z is equal to 0, but intuitively it should give 42
```
We are not going to fix it now. We will fix it together with #11.